### ISSUE TITLE
transformers.set_seed seems to do nothing

### ISSUE THREAD

**Issue by mojejmenojehonza** [huggingface/transformers#23215]
### System Info

- `transformers` version: 4.28.0.dev0
- Platform: Linux-5.19.0-41-generic-x86_64-with-glibc2.35
- Python version: 3.10.10
- Huggingface_hub version: 0.13.3
- Safetensors version: not installed
- PyTorch version (GPU?): 2.0.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@gante, @younesbelkada

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

1. install transformes with support for alpaca model
2. run this code with one seed
3. run this code with any other seed
4. see that the results are the same
```
from transformers import GenerationConfig, LlamaTokenizer, LlamaForCausalLM, set_seed
from torch import float16, compile, no_grad


set_seed(621)

# Enhances prompt
def enhance_prompt(prompt, input=None):
    if input:
        return f"""Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
{prompt}

### Input:
{input}

### Response:"""
    else:
        return f"""Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
{prompt}

### Response:"""

# Gets response from Alpaca
def get_response(prompt):
    with no_grad():
        outputs = alpaca.generate(input_ids=tokenizer(prompt, return_tensors="pt").input_ids.to("cuda"), generation_config=generation_config, return_dict_in_generate=True, output_scores=True)
    outputs = tokenizer.decode(outputs.sequences[0], skip_special_tokens=True)
    return outputs.split("### Response:")[1]


# Sets up Alpaca
tokenizer = LlamaTokenizer.from_pretrained("chainyo/alpaca-lora-7b")
alpaca = LlamaForCausalLM.from_pretrained("chainyo/alpaca-lora-7b", load_in_8bit=True, torch_dtype=float16, device_map="auto")
generation_config = GenerationConfig(temperature=0.2, top_p=0.75, top_k=40, num_beams=4, max_new_tokens=64)
alpaca.eval()
compile(alpaca)

# Gets output from Alpaca
prompt = enhance_prompt("Write a simple poem about flowers.")
out = get_response(prompt)

# Prints Alpaca's output
print(out)
```

### Expected behavior

Model will output two different answers but now it gives the same every seed I try.

**Comment by gante** [comment]
@mojejmenojehonza 👋

Two points:
1. You should pass `do_sample=True` in your generation config or in your `.generate()` call. Most models have it off by default, causing the generation to be deterministic (and ignoring parameters like `temperature`, `top_k`, etc).
2. With `temperature=0.2`, the relative weight of the most likely logits is massively increased, making generation almost deterministic. Even if there are no bugs in your script, it's far from guaranteed that two different seeds produce different outputs with such low temperature :)

**Comment by x6p2n9q8a4** [comment]
Yes!  I set the seed many times:   

1st:
from transformers import set_seed as transformers_set_seed
transformers_set_seed(3407)

2nd:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

3rd: 
trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    eval_dataset = val_dataset,
    dataset_text_field = "text",
    max_seq_length = max_seq_length,
    dataset_num_proc = 2,
    packing = False, # Can make training 5x faster for short sequences.
    args = SFTConfig(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        max_steps = 1500,
        learning_rate = 2e-4,
        fp16 = not is_bfloat16_supported(),
        bf16 = is_bfloat16_supported(),
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,...)

**Comment by AliHaider20** [comment]
@gante 
How do we set the seed for the retrieval and reranker models? 
I'm using OpenAI with RAG, but it is not working. Below is the code I used to set the seed.

```

from transformers import set_seed
SEED = 0
set_seed(SEED)

gpt = ChatOpenAI(
    model="gpt-4o",
    temperature=0,
    max_retries=2,
    max_tokens=4056,
    top_p=0.95,
    seed=SEED
)

```

**Comment by LysandreJik** [comment]
@AliHaider20 this is the `transformers` repo, not openai

**Comment by AliHaider20** [comment]
My issue is the same. Transformer seed is not working for me

**Comment by LysandreJik** [comment]
@AliHaider20 do you mind opening a new issue with your specific code and problem so that we may take a look? Thank you.

**Comment by AliHaider20** [comment]
Sure, I think that's the problem with OpenAI, and it can't be solved. If anyone gets to this page, just for reference:

[Model level controls for consistent outputs - seed and system_fingerprint](https://cookbook.openai.com/examples/reproducible_outputs_with_the_seed_parameter#model-level-controls-for-consistent-outputs---seed-and-system_fingerprint)
[seed](https://cookbook.openai.com/examples/reproducible_outputs_with_the_seed_parameter#seed)
If specified, our system will make a best effort to sample deterministically, such that repeated requests with the same seed and parameters should return the same result. Determinism is not guaranteed, and you should refer to the system_fingerprint response parameter to monitor changes in the backend.

[system_fingerprint](https://cookbook.openai.com/examples/reproducible_outputs_with_the_seed_parameter#system_fingerprint)
This fingerprint represents the backend configuration that the model runs with. It can be used in conjunction with the seed request parameter to understand when backend changes have been made that might impact determinism.This is the indicator on whether users should expect "almost always the same result".

**Comment by chrico-bu-uab** [comment]
> My issue is the same. Transformer seed is not working for me

They mean the `transformers` package, not transformers in general!
